flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/SavepointWindowReaderITCase.java

```diff
@@ -0,0 +1,370 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.api;
+
+import org.apache.flink.api.common.eventtime.WatermarkStrategy;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.runtime.state.StateBackend;
+import org.apache.flink.state.api.functions.WindowReaderFunction;
+import org.apache.flink.state.api.utils.AggregateSum;
+import org.apache.flink.state.api.utils.ReduceSum;
+import org.apache.flink.state.api.utils.SavepointTestBase;
+import org.apache.flink.streaming.api.TimeCharacteristic;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
+import org.apache.flink.streaming.api.functions.windowing.WindowFunction;
+import org.apache.flink.streaming.api.windowing.evictors.Evictor;
+import org.apache.flink.streaming.api.windowing.time.Time;
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
+import org.apache.flink.streaming.api.windowing.windows.Window;
+import org.apache.flink.streaming.runtime.operators.windowing.TimestampedValue;
+import org.apache.flink.util.Collector;
+
+import org.hamcrest.Matchers;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.List;
+
+/**
+ * IT Case for reading window operator state.
+ */
+public abstract class SavepointWindowReaderITCase<B extends StateBackend> extends SavepointTestBase {
+	private static final String uid = "stateful-operator";
+
+	private static final Integer[] numbers = { 1, 2, 3 };
+
+	protected abstract B getStateBackend();
+
+	@Test
+	public void testReduceWindowStateReader() throws Exception {
+		String savepointPath = takeSavepoint(numbers, source -> {
+			StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+			env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
+			env.setStateBackend(getStateBackend());
+			env.setParallelism(4);
+
+			env
+				.addSource(source)
+				.rebalance()
+				.assignTimestampsAndWatermarks(WatermarkStrategy
+					.<Integer>noWatermarks()
+					.withTimestampAssigner((event, timestamp) -> 0))
+				.keyBy(id -> id)
+				.timeWindow(Time.milliseconds(10))
+				.reduce(new ReduceSum())
+				.uid(uid)
+				.addSink(new DiscardingSink<>());
+
+			return env;
+		});
+
+		ExecutionEnvironment batchEnv = ExecutionEnvironment.getExecutionEnvironment();
+		ExistingSavepoint savepoint = Savepoint.load(batchEnv, savepointPath, getStateBackend());
+
+		List<Integer> results = savepoint
+			.timeWindow()
+			.reduce(uid, new ReduceSum(), Types.INT, Types.INT)
+			.collect();
+
+		Assert.assertThat("Unexpected results from keyed state", results, Matchers.containsInAnyOrder(numbers));
```

Comment:
Been trying to use this test as an example, to see if I can understand what's going on. Why would the savepoint contain all of the numbers? I would expect that at whatever point the savepoint is taken, each window pane would only contain the result of reducing whichever numbers have been processed so far. Are you assuming that the 3 numbers/keys are each distributed to a different parallel instance, and that's why all of them appear in the state?

Reply:
Yes, the number is the key and the value so they each go to a different window pane.